libc/string/string_tokpbrkstr.c

```diff
@@ -0,0 +1,875 @@
+	TEST_ASSERT_EQUAL_PTR(&loremStr[13], strtok(NULL, "r"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, sep));
+
+	/*
+	 * strtok() changes the content of a string,
+	 * that's why we have to set it once again
+	 */
+	memcpy(loremStr, LOREM_IPSUM, sizeof(LOREM_IPSUM));
+
+	/*
+	 * After passing all delimiters we cannot avoid misplaced tokens because
+	 * strtok looks for all elements in "ImDr" and breaks a string
+	 * in points where the first occurrence found
+	 */
+	TEST_ASSERT_EQUAL_PTR(loremStr, strtok(loremStr, sep));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[3], strtok(NULL, sep));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[5], strtok(NULL, sep));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[7], strtok(NULL, sep));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[11], strtok(NULL, sep));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[13], strtok(NULL, sep));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, sep));
+
+	/*
+	 * In this case, we tokenize the whole ASCII set using jump-over by
+	 * 2 elements. We are forced to jump over 2 places because strtok set
+	 *  divider placed as a null term and in this case, we are unable to
+	 * cover the whole set. Jumping over 2 places guarantees us output in the form
+	 * of one element form ascii set to do an assert with the place where
+	 * strtok has saved the stopping place
+	 */
+	TEST_ASSERT_EQUAL_PTR(&asciiStr[1], strtok(&asciiStr[1], "\2"));
+	for (i = 4; i < ASCII_LENGTH; i += 2) {
+		asciiSep[0] = i;
+		TEST_ASSERT_EQUAL_PTR(&asciiStr[i - 1], strtok(NULL, &asciiSep[0]));
+	}
+
+	/* Checking if set is empty */
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, &asciiSep[0]));
+
+	free(asciiStr);
+}
+
+
+TEST(string_tok, out_of_ascii)
+{
+	char divider[6] = { 0 },
+		 separator[ASCII_LENGTH],
+		 *extAsciiStr = create_extAscii_set();
+	int i;
+
+	divider[0] = extAsciiStr[0];
+	divider[1] = extAsciiStr[64];
+	divider[2] = extAsciiStr[128];
+	divider[3] = extAsciiStr[192];
+	divider[4] = extAsciiStr[254];
+
+
+	/* Checking ability for tokenize element above standard charset */
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[1], strtok(extAsciiStr, divider));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[65], strtok(NULL, divider));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[129], strtok(NULL, divider));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[193], strtok(NULL, divider));
+
+	/* 256 is not taken as a token because on element after it is a null term and it is recognized as the end of set */
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, divider));
+
+	free(extAsciiStr);
+	extAsciiStr = create_extAscii_set();
+	/*
+	 * Checking available tokenization on all elements
+	 * from outside of ASCII charset one by one, with
+	 * the first separator set as the first element of the set
+	 */
+	TEST_ASSERT_EQUAL_PTR(extAsciiStr, strtok(extAsciiStr, "\128"));
+	for (i = ASCII_LENGTH + 1; i < sizeof(separator); i++) {
+		separator[i - ASCII_LENGTH] = i;
+		TEST_ASSERT_EQUAL_PTR(&extAsciiStr[i - ASCII_LENGTH], strtok(NULL, &separator[i - ASCII_LENGTH]));
+	}
+
+	free(extAsciiStr);
+}
```

Comment:
We could also create a case, where the separator string contains the same character a few times, same with `str`, it could contain the same characters multiple times. There is no `big` case, we can do it at once in such one.

Reply:
Where is such case? I can't see it for `strtok()` function.

Reply:
Those cases are checked in `multi_call` in strtok and strtok_r